### DISCUSSION TITLE
ZHA Entity initialization

### DISCUSSION THREAD

**Discussion by Adminiuga** [zigpy/zigpy#582] [ZHA development]
Originally discussed in https://github.com/home-assistant/core/pull/44092#issuecomment-744049638

---

All very good questions, and perhaps question and answers should go into contributing guide

> 1. When a new device joins, we call the `async_configure_channel_specific()` to get configuration-specific attributes.

This is artifact from the past and we should not do this anymore.  The intent of the `async_configure()` method on the device is to 
1. Bind clusters() -- this is needed for attribute reporting (server clusters) and receiving commands from the "client clusters" -- remotes. Currently we do it for all clusters, but this should change in the future, as some devices limit the number of "bindings"
2. Configure attribute reporting -- to receive attribute updates

`async_configure_channel_specific()` is intended to do some extra specific configuration, e.g. reading groups from the LightLink cluster and add those groups to coordinator. But right now it serves mostly as back-compatibility, for entities which read attributes during configuration phase.
Previously, this (and `only_fromcache=True` flag to set initial state of the entity) was needed for some entities, like lights, because we had to have some attribute values known before we could create an entity with proper settings. 
Now, with ability to access zigpy attribute cache directly and that `async_initialize()` is called on the device for all channels before the entities are added to hass you should do all the attribute reads in `async_initialize_channel_specific()`. Keep in mind, all attributes defined for attribute reporting, are read automatically in the base channel class, so in initialize channel specific you should read only additional attributes. 
In short:
In current architecture, before an entity of the a device is added, the following are ran:
- When a new device joins or previous device rejoins:
  1. `async_configure() and `async_configure_channel_specific()`
  2. `async_initialize()` and `async_initialize_channels_specific()`
- When HA is restarted:
  1. `async_initialize()` and `async_initialize_channels_specific()` -- for battery operated devices we do `from_cache=True` and `only_from_cache=True`

Whenever you need `only_from_cache=True` should be replaced with `cluster.get["attribute_value"]`.

> Shortly after that the ZhaNumber.async_update() is called that also fetches the present_value attribute. Even if I add present_value to async_configure_channel_specific, that's two calls to get attributes instead of one so assert cluster.read_attributes.call_count == 1 in the test fails.

When we add entities, we do `update_before_add=True` which calls `async_update()` and which may call `read_attributes()`. So
you end up with two calls to  `read_attributes()`:
1. from `channel.async_initialize()`
2. from `async_update()` from `update_before_add=True`

This is another artifact from backward compatibility and I feel that  `update_before_add=True` is not needed anymore. It was calling `async_update()` with `only_from_cache=True` to set initial state of the entity. Should be replaced `cluster.get("attr_name")` instead and `update_before_add` removed.

Entities still should/may have `async_update()` method, for `homeassistant.update_entity` service call to work

> Should I remove the async_update at all or it could be useful? Is it OK to read attributes twice? Should I test for cluster.read_attributes.call_count at all? Strangely, it does not happen with cover entities and I can't figure out the difference.

make it so it does not do `update_before_add`. Currently `sensor` zha entity works without it. My plan is to remove for all, just don't have time.

**Comment by pipiche38** [comment]
interesting, but what is the conclusion ? to be used or not ? 

Still for me the quirks have the device specific informations and the bind+configure reporting should be done at that level, otherwise this will enforce the application layer to ALSO keep track of the device specific things.

**Reply by puddly** [reply]
This is an interesting question.

I think zigpy should be fully configuring joining devices. Most people who want to use zigpy outside of HA (if they can dig up an example somewhere) realize that they are responsible for initializing every joining device, which I think can easily be handled by zigpy itself with hooks to allow the higher level application to provide per-device settings, its own defaults, or to replace zigpy's initialization entirely.

Binding and attribute reporting settings can be stored in the zigpy appdb as columns on the clusters and attribute tables, basically extending the current initialization process with a second stage and making it more fault-tolerant. That information can be cleared if a device re-joins with a new NWK, preventing partially-initialized devices from being exposed to the user and hopefully making the database a proper reflection of a device's actual state.

**Reply by pipiche38** [reply]
@puddly that would indeed makes zigpy much powerfull than only been an abstract layer to the radio devices.
That means that the quirks will also have to integrate that as well

Now if I understood this is not yet the case, and if I want to start I will have to do the bind and reporting ?

**Reply by pipiche38** [reply]
a side remark I believe zigpy is fully initialising IAS devices and in such do the full registration ?

**Reply by Adminiuga** [reply]
>I think zigpy should be fully configuring joining devices. Most people who want to use zigpy outside of HA (if they can dig up an example somewhere) realize that they are responsible for initializing every joining device

I disagree on that one. `zigpy` should focus on handling the Zigbee communication only.  But to your point, a simpler "application" above zigpy handling the basic functionality could be useful and it still should be separate abstraction layer from `zigpy`

> Binding and attribute reporting settings can be stored in the zigpy appdb as columns on the clusters and attribute tables, basically extending the current initialization process with a second stage and making it more fault-tolerant. That information can be cleared if a device re-joins with a new NWK, preventing partially-initialized devices from being exposed to the user and hopefully making the database a proper reflection of a device's actual state

The challenge is to keep this kind of information in sync (especially for end devices or manufacturers which really don't respond to request on demand) and if you clear it up on the new NWK, you might as well reconfigure the bindings/reports again 🤷

**Reply by pipiche38** [reply]
Fair, but still where is quirks ? I would consider that is the purpose of quirks to do so, as it does already a lot of things which are not radio related and so forth are not zigpy .
The zhaquirks got all manufacturer specifics like the Xiaomi , like the Tuya and the Configure Reporting as well as the Bind must be there.
Otherwise I don't see the fact to have one more layer doing it

**Reply by MattWestb** [reply]
One IKEA light and remote is working "out of the box" with ZHA / Zigpy without quirks installed / disabled only that the remote is reporting half of its battery (IKEA have making it  wrong).

Quirk handling is only fixing band / not zigbee standard things in devices and its also must do some very low and also high level things like tuya is not having normal clusters but the quirk is "making" them and also need doing more things for it to working.

But the host system over Zigpy is only understanding zigbbe commands and the radio libs is the interface to the radio stack and the quirk is between for fixing "the bad things" but its little more advanced then only fixing bad battery reporting .

**Reply by pipiche38** [reply]
@MattWestb I do not understand your Ikea example , You have to Bind and set the configure reporting if you want to get the Led status if operated by a remote or a group command.
As the battery I'm using a 5 buttons and I get the battery correctly, but did set bind + configure reporting.

So if you then consider that Bind+Configure reporting which are manufacturer specific shouldn't be done by either zigpy or quirk, then I take the point, but consider that you are missing an important layer and when I see what Quirk,is doing especially for Tuya , I think there is miss-match somewhere

**Reply by Hedda** [reply]
> This is an interesting question.
> 
> I think zigpy should be fully configuring joining devices. Most people who want to use zigpy outside of HA (if they can dig up an example somewhere) realize that they are responsible for initializing every joining device, which I think can easily be handled by zigpy itself with hooks to allow the higher level application to provide per-device settings, its own defaults, or to replace zigpy's initialization entirely.
> 
> Binding and attribute reporting settings can be stored in the zigpy appdb as columns on the clusters and attribute tables, basically extending the current initialization process with a second stage and making it more fault-tolerant. That information can be cleared if a device re-joins with a new NWK, preventing partially-initialized devices from being exposed to the user and hopefully making the database a proper reflection of a device's actual state.

@zoic21 if you do not mind me asking, what is your opinions on this from working on zigpy with [new Jeedom Zigbee Plugin](https://github.com/zigpy/zigpy/discussions/725)?

**Reply by MattWestb** [reply]
The binding and configure attribute reporting is being done from the upper system with normal zigbee commands. Zigpy / radiolib is not doing it normally.

The 5 button remote is reporting 100% battery but its also reporting 2 batteries = 100 / 2 = 50% The quirk is over riding it by doing * 2 and you is getting 100% battery.

Some devices is only reporting battery voltage and the quirk is calculating the % so the upper system is getting it ala zigbee standard.

One tuya TRY is after the quirk one "normal zigbee TRV" = have all the cluster that one trv shall having by the zigbee standard and can being controlled as one TRV from the upper system.
Without the quirk the upper system cant controlling it then its not understanding normal zigbee commands.

One tuya switch that  is sending tuya commands also the quirk is making zigbee clusters so the upper system can understanding and using it as one normal zigbee switch device.

Then doing all this magic some workarounds can being needed for doing the config and reporting in the quirk but  normal its all is being made in thee upper system.

Therefor its better / easier using one "normal" device to working OK in the upper system and not starting with the "advances" that have much magic that can making problem all the way from the radio lib until the upper system.

**Comment by Hedda** [comment]
Side-note; I think best would be to avoid developing with Aqara/Xiaomi and Tuya devices in the very beginning of a proof-of-concept.

I think we can not really stress this enough; you should really just avoid all Aquara/Xiaomi and Tuya devices (or any other generic China/Chinese Zigbee device) as they are almost all infamously known to not follow Zigbee specifications and as such will for sure cause guaranteed headaches unnecessary during early troubleshooting. 

That follows the "low-hanging fruit" principle to gain momentum and not getting stuck by making it much harder for yourself than it really has to be in the right in the beginning. Instead, suggest focus on getting easier devices to work with and simpler functions working first.

Others might have different suggestions here but I believe the general recommendation is to start out to with IKEA Tradfri (IKEA Trådfri) series devices instead of to begin with as they are better known to follow standard Zigbee specifications without deviating too much.

**Reply by MattWestb** [reply]
110% agree with @Hedda !!

Dont start with Xiaomi motion sensor then its not one "real zigbee device" is not reporting on normal cluster and is not doing anything as zigbee device shall doing.

One IKEA motion sensor is 95% zigbee compatible (battery is wrong) and it technically one on/of controller that is sending on with time off in x seconds for controlling  lights.

First basic functions like attribute reporting and binding working you is have 90% of all devices working and the last 10% is have much magic that need much time to getting working in the upper system.

**Reply by pipiche38** [reply]
I share your point, but this represent a great portion of the market. On the other side I'm also testing with Legrand Netatmo and for me the main point if the Binding/Configure reporting topic.

An my main question is to understand in which layer the quirks is in the overall architecture design, as I see quirks actions happening at various level, and that is why I'm questioning to have the Quirks doing the Bind+CfgReporting

**Reply by pipiche38** [reply]
last @MattWestb to introduce myself I'm developping a Zigbee plugin for Domoticz since more than 3 years , so I know pretty much what all of those devices behaves and works. I don't understand @hedda and you answers as regards to Xiami motion.
Let stick to the purpose of this discussion which is Zigbee device initialisation end to end ( as @puddly was mentioning )

**Reply by MattWestb** [reply]
Devices that is OK in zigbee point of view is not needing any quirk and is getting the binding and reporting from the upper system without problems and is also working if the quirk is not installed / disabled.

If you is taking one tuya TRV that is not having all the clusters the quirk is making them and also some parts of binding and reporting between the device and the upper system so its working as one "normal" zigbee device from the upper system. You can saying the devices is bonded and reporting the the quirk so the quirk can acting as one normal device and being configurated as on normal device therfor the quirk need doing binding and reporting to the "problematic" device so it can doing its magic and deliver normal zigbee cluster  to the upper system.

And i think(=not knowing) the quirk is working in many layers for doing the "converting" devices to zigbee devices with complete different property (tuya TRV is on Zigbee light switch or plug and after the quirk is one Zigbee TRV).

**Reply by MattWestb** [reply]
I think only @puddly  and @Adminiuga  can helping you in the deep technical things.

Wath @Hedda  and i is saying is dont digging in the quirk at the first stage.
Getting the basic working and then going for the quirk things or you is getting stuck in very deep problems that you cant getting solved without learning all code that zigpy and its radio libs is s using.

**Reply by pipiche38** [reply]
@MattWestb my topic is to have a end to end provisioning done. Today I do understand that it is done by " the upper system", which is in your case HA ?

**Reply by pipiche38** [reply]
> I think only @puddly and @Adminiuga can helping you in the deep technical things.
> 
> Wath @Hedda and i is saying is dont digging in the quirk at the first stage.
> Getting the basic working and then going for the quirk things or you is getting stuck in very deep problems that you cant getting solved without learning all code that zigpy and its radio libs is s using.

My point is to understand what zigpy , quirks does in order to architect the plugin for Domotciz in the right manner. I'm also challenging that fact that if this is not quirks that an extra layer is lacking.

**Reply by MattWestb** [reply]
Im hardware / telecom engineer and not software dev but have much experience of communication systems and is for the moment using ZHA in production and more for test (and deCONZ as x production and reference system).

And yes Zigpy is one lib for host system talking zigbee with devices with different radios.
The host system is only talking zigbee to the lib and is doing all the upper things then the zigpy is only one "converter" for commands.

**Comment by pipiche38** [comment]
I'm sorry but I still do not understand the all stacks. 
Ok so zigpy take care of zcl and zdo communication layer, and I'm fully aligned with that.
On the other side, when I do the pairing of a device whatever brand it is, the zigpy + quirks layers do a full device interview to get 👍 - profileId
DeviceId
Cluster In
Cluster Out

That is not a communication stuff, but a real device interview.

So I feel that we are mixing the zigpy initial idea which is an abstract layer of the Radio controlers, and additional stuff which have been implemented on top of this communication layer.

And that is my point, why an hosting application seating on this set of libs (zigpy + quirks) doing especially the full device discovery is not perfoming the bind + configure reporting set which is basically what is not done to get a full working device .
and that is were I'm in line with @puddly  statement 

> I think zigpy should be fully configuring joining devices. Most people who want to use zigpy outside of HA (if they can dig up an example somewhere) realize that they are responsible for initializing every joining device, which I think can easily be handled by zigpy itself with hooks to allow the higher level application to provide per-device settings, its own defaults, or to replace zigpy's initialization entirely.

[EDITED] last, let's imagine that you have a non-standard device, via the zhaquirks it will be presented as a standard with the standard cluster and attributes. However if bind+cfgrpt have to be performed the host will make it on which cluster ( the standard or the non-standard)
